token/transfer-hook/interface/src/onchain.rs

```diff
@@ -92,3 +96,514 @@ pub fn add_cpi_accounts_for_execute<'a>(
         .push(AccountMeta::new_readonly(*program_id, false));
     Ok(())
 }
+
+/// Helper to add accounts required for an `ExecuteInstruction` on-chain,
+/// looking through the additional account infos to add the proper accounts.
+///
+/// Note this helper is designed to add the extra accounts that will be
+/// required for a CPI to a transfer hook program. However, the instruction
+/// being provided to this helper is for the program that will CPI to the
+/// transfer hook program. Because of this, we must resolve the extra accounts
+/// for the `ExecuteInstruction` CPI, then add those extra resolved accounts to
+/// the provided instruction.
+#[allow(clippy::too_many_arguments)]
+pub fn add_extra_accounts_for_execute_cpi<'a>(
+    cpi_instruction: &mut Instruction,
+    cpi_account_infos: &mut Vec<AccountInfo<'a>>,
+    program_id: &Pubkey,
+    source_pubkey: &Pubkey,
+    mint_pubkey: &Pubkey,
+    destination_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
+    amount: u64,
+    additional_accounts: &[AccountInfo<'a>],
+) -> ProgramResult {
+    let extract_account_info = |key: &Pubkey, infos: &[AccountInfo<'a>]| {
+        infos
+            .iter()
+            .find(|&info| info.key == key)
+            .ok_or(TransferHookError::IncorrectAccount)
+            .map(|info| info.clone())
+    };
+
+    let validate_state_pubkey = get_extra_account_metas_address(mint_pubkey, program_id);
+
+    let source_info = extract_account_info(source_pubkey, &cpi_account_infos)?;
+    let mint_info = extract_account_info(mint_pubkey, cpi_account_infos)?;
+    let destination_info = extract_account_info(destination_pubkey, cpi_account_infos)?;
+    let authority_info = extract_account_info(authority_pubkey, cpi_account_infos)?;
+    let program_info = extract_account_info(program_id, additional_accounts)?;
+    let validate_state_info = extract_account_info(&validate_state_pubkey, additional_accounts)?;
+
+    // Check to make sure the provided keys are in the instruction
+    if [
+        source_pubkey,
+        mint_pubkey,
+        destination_pubkey,
+        authority_pubkey,
+    ]
+    .iter()
+    .any(|&key| {
+        !cpi_instruction
+            .accounts
+            .iter()
+            .any(|meta| meta.pubkey == *key)
+    }) {
+        Err(TransferHookError::IncorrectAccount)?;
+    }
```

Comment:
I know we have this in the offchain helper, but I don't know if this check is worth it on chain, since it does introduce a lot of linear searches, which burns compute units.

I'd remove it here, but I'll leave it up to you

Reply:
I think the check is useful, however the error really isn't 💀, so I'm not against taking it out.

What about something like this?

```rust
// Check to make sure the provided keys are in the instruction
let mut required_keys = std::collections::HashSet::from([
    source_pubkey,
    mint_pubkey,
    destination_pubkey,
    authority_pubkey,
]);
for account_info in cpi_account_infos {
    required_keys.remove(&account_info.key);
    if required_keys.is_empty() {
        break;
    }
}
if !required_keys.is_empty() {
    Err(TransferHookError::IncorrectAccount)?;
}
```

Reply:
I think this is less relevant considering the changes from the other comment. Removed.